src/internal/operators/mergeMap.ts

```diff
@@ -74,93 +86,125 @@ export function mergeMap<T, R, O extends ObservableInput<any>>(
+            return;
+          }
+
+          // Subscribe to the inner source
+          active++;
+          let innerSubs: Subscription;
+          subscriber.add(
+            (innerSubs = innerSource.subscribe(
+              new MergeMapSubscriber(
+                subscriber,
+                (innerValue) => {
+                  // INNER SOURCE NEXT
+                  // We got a value from the inner source, emit it from the result.
+                  subscriber.next(innerValue);
+                },
+                () => {
+                  // INNER SOURCE COMPLETE
+                  // Decrement the active count to ensure that the next time
+                  // we try to call `doInnerSub`, the number is accurate.
+                  active--;
+                  if (buffer.length > 0) {
+                    // If we have more values in the buffer, try to process those
+                    // Note that this call will increment `active` ahead of the
+                    // next conditional, if there were any more inner subscriptions
+                    // to start.
+                    doInnerSub();
+                  }
+                  if (isComplete && active === 0) {
+                    // If the outer is complete, and there are no more active,
+                    // then we can complete the resulting observable subscription
+                    subscriber.complete();
+                  }
+                  // Make sure to teardown the inner subscription ASAP.
+                  innerSubs?.unsubscribe();
+                }
+              )
+            ))
+          );
+        }
+      };
+
+      let outerSubs: Subscription;
+      outerSubs = source.subscribe(
+        new MergeMapSubscriber(
+          subscriber,
+          (value) => {
+            // OUTER SOURCE NEXT
+            // Push the value onto the buffer. We have no idea what the concurrency limit
+            // is and we don't care. Just buffer it and then call `doInnerSub()` to try to
+            // process what is in the buffer.
+            buffer.push(value);
+            doInnerSub();
+          },
+          () => {
+            // OUTER SOURCE COMPLETE
+            // We don't necessarily stop here. If have any pending inner subscriptions
+            // we need to wait for those to be done first. That includes buffered inners
+            // that we haven't even subscribed to yet.
+            isComplete = true;
+            if (active === 0 && buffer.length === 0) {
+              // Nothing is active, and nothing in the buffer, with no hope of getting any more
+              // we can complete the result
+              subscriber.complete();
+            }
+            // Be sure to teardown the outer subscription ASAP, in any case.
+            outerSubs?.unsubscribe();
+          }
+        )
+      );
+    });
 }
 
+// TODO(benlesh): This may end up being so common that we can centralize on one Subscriber for a few operators.
```

Comment:
Yes. This is one of the perf-related issues I'll be looking at in a week or so. If we can eliminate derived subscribers, a whole class of performance issues will simply disappear.